go/vt/vttablet/tabletmanager/vreplication/vplayer.go

```diff
@@ -0,0 +1,632 @@
+		if err == nil || err == io.EOF {
+			return errors.New("vstream ended")
+		}
+		return err
+	}
+}
+
+func (vp *vplayer) applyEvents(ctx context.Context, relay *relayLog) error {
+	for {
+		items, err := relay.Fetch()
+		if err != nil {
+			return err
+		}
+		// This covers two situations:
+		// 1. Fetch was idle for idleTimeout.
+		// 2. We've been receiving empty events for longer than idleTimeout.
+		// In both cases, now > timeLastSaved. If so, any unsaved GTID should be saved.
+		if time.Now().Sub(vp.timeLastSaved) >= idleTimeout && vp.unsavedGTID != nil {
+			// Although unlikely, we should not save if a transaction is still open.
+			// This can happen if a large transaction is split as multiple events.
+			if !vp.dbClient.InTransaction {
+				if err := vp.updatePos(vp.unsavedGTID.Timestamp); err != nil {
+					return err
+				}
+			}
+		}
+		for i, events := range items {
+			for j, event := range events {
+				mustSave := false
+				switch event.Type {
+				case binlogdatapb.VEventType_COMMIT:
+					if vp.pos.Equal(vp.stopPos) {
+						mustSave = true
+						break
+					}
+					if hasAnotherCommit(items, i, j+1) {
+						continue
+					}
+				}
+				if err := vp.applyEvent(ctx, event, mustSave); err != nil {
+					return err
+				}
+			}
+		}
+	}
+}
+
+func hasAnotherCommit(items [][]*binlogdatapb.VEvent, i, j int) bool {
+	for i < len(items) {
+		for j < len(items[i]) {
+			// We ignore GTID, BEGIN, FIELD and ROW.
+			switch items[i][j].Type {
+			case binlogdatapb.VEventType_COMMIT:
+				return true
+			case binlogdatapb.VEventType_DDL:
+				return false
+			}
+			j++
+		}
+		j = 0
+		i++
+	}
+	return false
+}
+
+func (vp *vplayer) applyEvent(ctx context.Context, event *binlogdatapb.VEvent, mustSave bool) error {
+	switch event.Type {
+	case binlogdatapb.VEventType_GTID:
+		pos, err := mysql.DecodePosition(event.Gtid)
+		if err != nil {
+			return err
+		}
+		vp.pos = pos
+		vp.unsavedGTID = event
+		if vp.stopPos.IsZero() {
+			return nil
+		}
+		if !vp.pos.Equal(vp.stopPos) && vp.pos.AtLeast(vp.stopPos) {
+			// Code is unreachable, but bad data can cause this to happen.
+			if err := vp.setState(binlogplayer.BlpStopped, fmt.Sprintf("next event position %v exceeds stop pos %v, exiting without applying", vp.pos, vp.stopPos)); err != nil {
+				return err
+			}
+			return io.EOF
+		}
+	case binlogdatapb.VEventType_BEGIN:
+		// No-op: begin is called as needed.
+	case binlogdatapb.VEventType_COMMIT:
+		if mustSave {
+			if err := vp.dbClient.Begin(); err != nil {
+				return err
+			}
+		}
+
+		if !vp.dbClient.InTransaction {
+			return nil
+		}
+		if err := vp.updatePos(event.Timestamp); err != nil {
+			return err
+		}
+		posReached := !vp.stopPos.IsZero() && vp.pos.Equal(vp.stopPos)
+		if posReached {
+			if err := vp.setState(binlogplayer.BlpStopped, fmt.Sprintf("Stopped at position %v", vp.stopPos)); err != nil {
+				return err
+			}
+		}
+		if err := vp.dbClient.Commit(); err != nil {
```

Comment:
If the commit fails, the retry will eventually reach the position. So, we're fine. We can also do this as a separate commit, but only later. But I think it's better if we also guarantee that we atomically stop on reaching the stop position. Otherwise, we'll be in running state even after reaching stop pos, which feels weird.

Doing it before is definitely not good because if the commit of the event subsequently fails, we'll be forever in stopped state and will never reach the stop pos.